src/librbd/Group.cc

```diff
@@ -379,4 +382,405 @@ int image_get_group(ImageCtx *ictx, group_spec_t *group_spec)
+	       << dendl;
+    return r;
+  }
+  string group_header_oid = util::group_header_name(group_id);
+
+  std::vector<group_image_status_t> images;
+  r = librbd::group_image_list(group_ioctx, group_name, images);
+  if (r < 0) {
+    return r;
+  }
+  int n = images.size();
+  std::vector<librbd::IoCtx*> io_ctxs;
+  std::vector<librbd::ImageCtx*> ictxs;
+  std::vector<C_SaferCond*> on_finishes;
+  for (auto i: images) {
+    librbd::IoCtx* image_io_ctx = new librbd::IoCtx;
+
+    r = rados.ioctx_create2(i.pool, *image_io_ctx);
+    if (r < 0) {
+      ldout(cct, 1) << "Failed to create io context for image" << dendl;
+    }
+
+    librbd::ImageCtx* image_ctx = new ImageCtx(i.name.c_str(), "", "",
+					       *image_io_ctx, false);
+
+    C_SaferCond* on_finish = new C_SaferCond;
+
+    image_ctx->state->open(on_finish);
+
+    ictxs.push_back(image_ctx);
+    on_finishes.push_back(on_finish);
+  }
+  int ret_code = 0;
+  for (int i = 0; i < n; ++i) {
+    r = on_finishes[i]->wait();
+    delete on_finishes[i];
+    if (r < 0) {
+      delete ictxs[i];
+      ictxs[i] = nullptr;
+      ret_code = r;
+    }
+  }
+  if (ret_code != 0) {
+    goto cleanup;
+  }
+  for (auto i: ictxs) {
+    i->exclusive_lock->block_requests(-EBUSY);
+  }
+  for (int i = 0; i < n; ++i) {
+    ImageCtx *ictx = ictxs[i];
+    RWLock::RLocker owner_lock(ictx->owner_lock);
+
+    on_finishes[i] = new C_SaferCond;
+    ictx->exclusive_lock->request_lock(on_finishes[i]);
+  }
+
+  ret_code = 0;
+  for (auto i: on_finishes) {
+    r = i->wait();
+    delete i;
+    if (r < 0) {
+      ret_code = r;
+    }
+  }
+  if (ret_code != 0) {
+    goto cleanup;
+  }
+
+  r = cls_client::group_snap_next_seq(&group_ioctx, group_header_oid, &snap_seq);
+  if (r < 0) {
+    ret_code = r;
+    goto cleanup;
+  }
+
+  gs.id = snap_seq;
+  gs.uuid = "";
+  gs.name = string(snap_name);
+  gs.state = cls::rbd::GROUP_SNAPSHOT_STATE_PENDING;
+
+  r = cls_client::group_snap_save(&group_ioctx, group_header_oid, gs);
+  if (r < 0) {
+    ret_code = r;
+    goto cleanup;
+  }
+
+  image_snaps = vector<cls::rbd::ImageSnapshotRef>(n, cls::rbd::ImageSnapshotRef());
+
+  ind_snap_name = calc_ind_image_snap_name(group_ioctx.get_id(),
+					   group_id,
+					   snap_seq);
+
+  for (int i = 0; i < n; ++i) {
+    ImageCtx *ictx = ictxs[i];
+    cls::rbd::SnapshotNamespace ne =
+			 cls::rbd::GroupSnapshotNamespace(group_ioctx.get_id(),
+							  group_id,
+							  snap_seq);
+
+    C_SaferCond* on_finish = new C_SaferCond;
+
+    ictx->operations->snap_create(ind_snap_name.c_str(), ne, on_finish);
+
+    on_finishes[i] = on_finish;
+  }
+
+  ret_code = 0;
+  for (int i = 0; i < n; ++i) {
+    r = on_finishes[i]->wait();
+    delete on_finishes[i];
+    if (r < 0) {
+      ret_code = r;
+    } else {
+      ImageCtx *ictx = ictxs[i];
+      ldout(cct, 1) << "Get snap id with name " << ind_snap_name << dendl;
+      ictx->snap_lock.get_read();
+      snap_t snap_id = ictx->get_snap_id(ind_snap_name);
```

Comment:
If this double-lookup is required, I'd rather have the SnapshotCreateRequest state machine take an optional snap_id pointer to save the assigned snap id into before completing to avoid this lookup.